<a href="https://colab.research.google.com/github/G3ntl3g1ant/sales-predictions/blob/main/Project_1_Part_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Sales Predictions
Kai Dawson-Fischer

## Loading Data

In [3]:
import pandas as pd

In [4]:
filename = '/content/drive/MyDrive/Coding Dojo/Data Spreadsheets/sales_predictions.csv'
# sp = sales predictions data frame
sp = pd.read_csv(filename)
sp.head()

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
0           FDA15         9.30          Low Fat         0.016047   
1           DRC01         5.92          Regular         0.019278   
2           FDN15        17.50          Low Fat         0.016760   
3           FDX07        19.20          Regular         0.000000   
4           NCD19         8.93          Low Fat         0.000000   

               Item_Type  Item_MRP Outlet_Identifier  \
0                  Dairy  249.8092            OUT049   
1            Soft Drinks   48.2692            OUT018   
2                   Meat  141.6180            OUT049   
3  Fruits and Vegetables  182.0950            OUT010   
4              Household   53.8614            OUT013   

   Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
0                       1999      Medium               Tier 1   
1                       2009      Medium               Tier 3   
2                       1999      Medium               Tier 1   
3                       1998         NaN               Tier 3   
4                       1987        High               Tier 3   

         Outlet_Type  Item_Outlet_Sales  
0  Supermarket Type1          3735.1380  
1  Supermarket Type2           443.4228  
2  Supermarket Type1          2097.2700  
3      Grocery Store           732.3800  
4  Supermarket Type1           994.7052

In [5]:
sp.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8523 entries, 0 to 8522
Data columns (total 12 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Item_Identifier            8523 non-null   object 
 1   Item_Weight                7060 non-null   float64
 2   Item_Fat_Content           8523 non-null   object 
 3   Item_Visibility            8523 non-null   float64
 4   Item_Type                  8523 non-null   object 
 5   Item_MRP                   8523 non-null   float64
 6   Outlet_Identifier          8523 non-null   object 
 7   Outlet_Establishment_Year  8523 non-null   int64  
 8   Outlet_Size                6113 non-null   object 
 9   Outlet_Location_Type       8523 non-null   object 
 10  Outlet_Type                8523 non-null   object 
 11  Item_Outlet_Sales          8523 non-null   float64
dtypes: float64(4), int64(1), object(7)
memory usage: 799.2+ KB


## Data Cleaning

### First to identify the infomation of the data frame itself.

1) There are **8523** Rows and **11** Columns for this Data Frame.


2) As far as specific datatypes, there are ***integers***, ***floats*** and ***objects***.

In [6]:
sp.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8523 entries, 0 to 8522
Data columns (total 12 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Item_Identifier            8523 non-null   object 
 1   Item_Weight                7060 non-null   float64
 2   Item_Fat_Content           8523 non-null   object 
 3   Item_Visibility            8523 non-null   float64
 4   Item_Type                  8523 non-null   object 
 5   Item_MRP                   8523 non-null   float64
 6   Outlet_Identifier          8523 non-null   object 
 7   Outlet_Establishment_Year  8523 non-null   int64  
 8   Outlet_Size                6113 non-null   object 
 9   Outlet_Location_Type       8523 non-null   object 
 10  Outlet_Type                8523 non-null   object 
 11  Item_Outlet_Sales          8523 non-null   float64
dtypes: float64(4), int64(1), object(7)
memory usage: 799.2+ KB


#### 3) Second, the data frame needs to be checked for any duplications

Currently there are no duplicates in this data frame.

In [7]:
sp.duplicated().sum()

0

Searching through any data frames, the first thing that must be identified is if there is any missing data, to begin:

To help organize and identify the data, the index was changed to **Item_Identifier** since each item has a ***unique identifier***.

In [8]:
sp2 = pd.read_csv(filename)
sp2.set_index('Item_Identifier', inplace=True)
sp2.head(2)

Item_Weight Item_Fat_Content  Item_Visibility    Item_Type  \
Item_Identifier                                                               
FDA15                   9.30          Low Fat         0.016047        Dairy   
DRC01                   5.92          Regular         0.019278  Soft Drinks   

                 Item_MRP Outlet_Identifier  Outlet_Establishment_Year  \
Item_Identifier                                                          
FDA15            249.8092            OUT049                       1999   
DRC01             48.2692            OUT018                       2009   

                Outlet_Size Outlet_Location_Type        Outlet_Type  \
Item_Identifier                                                       
FDA15                Medium               Tier 1  Supermarket Type1   
DRC01                Medium               Tier 3  Supermarket Type2   

                 Item_Outlet_Sales  
Item_Identifier                     
FDA15                    3735.1380  
DRC01                     443.4228

4) Now to see if there are any missing values. 

In [9]:
sp2.isna().sum()

Item_Weight                  1463
Item_Fat_Content                0
Item_Visibility                 0
Item_Type                       0
Item_MRP                        0
Outlet_Identifier               0
Outlet_Establishment_Year       0
Outlet_Size                  2410
Outlet_Location_Type            0
Outlet_Type                     0
Item_Outlet_Sales               0
dtype: int64

#### 5) First there is missing data from the **'Item_Weight'** column, because there are so many entries missing and assuming the weight of these items could possibly taint the results, for now **'Missing'** will be the replacement variable until there is data to replace the missing information, or if it is found to be unnecessary. 

In [10]:
# What values need to be changed
sp2['Item_Weight'].fillna('Missing', inplace = True)
# Checking to see what other items are missing
sp2.isna().sum()

Item_Weight                     0
Item_Fat_Content                0
Item_Visibility                 0
Item_Type                       0
Item_MRP                        0
Outlet_Identifier               0
Outlet_Establishment_Year       0
Outlet_Size                  2410
Outlet_Location_Type            0
Outlet_Type                     0
Item_Outlet_Sales               0
dtype: int64

Now all that is left is the **'Outlet_Size'** column to fix, first the values for **'Outlet_Size'** must be identified

In [11]:
# Checking the values of the 'Outlet_Size' column
sp2['Outlet_Size'].value_counts()

Medium    2793
Small     2388
High       932
Name: Outlet_Size, dtype: int64

Now that the values are stated we can plug in the average outlet size to fill in the rest of the missing data. 

First, the values are changed to an int.

In [12]:
# Values to be changed
replace_values = {'Small' : 1, 'Medium' :2, 'High': 3}
sp2['Outlet_Size'] = sp2['Outlet_Size'].replace(replace_values)
sp2['Outlet_Size'].value_counts()

2.0    2793
1.0    2388
3.0     932
Name: Outlet_Size, dtype: int64

Then, to change the missing data into the **average** for **'Outlet_Size'**

In [13]:
# Now to look for the median of the oulet sizes
# mos = median outlet size
mos = sp2['Outlet_Size'].median()
sp2['Outlet_Size'].fillna(mos, inplace=True)

And finally, change into an int.

In [14]:
sp2['Outlet_Size'] = sp2['Outlet_Size'].astype(int)
sp2.info()

<class 'pandas.core.frame.DataFrame'>
Index: 8523 entries, FDA15 to DRG01
Data columns (total 11 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Item_Weight                8523 non-null   object 
 1   Item_Fat_Content           8523 non-null   object 
 2   Item_Visibility            8523 non-null   float64
 3   Item_Type                  8523 non-null   object 
 4   Item_MRP                   8523 non-null   float64
 5   Outlet_Identifier          8523 non-null   object 
 6   Outlet_Establishment_Year  8523 non-null   int64  
 7   Outlet_Size                8523 non-null   int64  
 8   Outlet_Location_Type       8523 non-null   object 
 9   Outlet_Type                8523 non-null   object 
 10  Item_Outlet_Sales          8523 non-null   float64
dtypes: float64(3), int64(2), object(6)
memory usage: 799.0+ KB


#### 6) Now to check the data frame once again for any missing values.

In [15]:
sp2.isna().sum()

Item_Weight                  0
Item_Fat_Content             0
Item_Visibility              0
Item_Type                    0
Item_MRP                     0
Outlet_Identifier            0
Outlet_Establishment_Year    0
Outlet_Size                  0
Outlet_Location_Type         0
Outlet_Type                  0
Item_Outlet_Sales            0
dtype: int64

#### Now with that out of the way, to check for needed columns and what data might be unnecessary for the data frame/data collection

In [16]:
sp2.nunique()

Item_Weight                   416
Item_Fat_Content                5
Item_Visibility              7880
Item_Type                      16
Item_MRP                     5938
Outlet_Identifier              10
Outlet_Establishment_Year       9
Outlet_Size                     3
Outlet_Location_Type            3
Outlet_Type                     4
Item_Outlet_Sales            3493
dtype: int64

After looking at the data as well as the key to the data, the **'Outlet_Establishment_Year'** and the **'Item_Fat_Content'** columns are unnecessary to the information we are looking for. So, they shall be removed.

In [17]:
sp2.drop(columns = {'Outlet_Establishment_Year', 'Item_Fat_Content'}, inplace=True)
sp2.head(2)

Item_Weight  Item_Visibility    Item_Type  Item_MRP  \
Item_Identifier                                                       
FDA15                   9.3         0.016047        Dairy  249.8092   
DRC01                  5.92         0.019278  Soft Drinks   48.2692   

                Outlet_Identifier  Outlet_Size Outlet_Location_Type  \
Item_Identifier                                                       
FDA15                      OUT049            2               Tier 1   
DRC01                      OUT018            2               Tier 3   

                       Outlet_Type  Item_Outlet_Sales  
Item_Identifier                                        
FDA15            Supermarket Type1          3735.1380  
DRC01            Supermarket Type2           443.4228

7) Finally, the break down of the finished data frame, including the breakdown for each numerical column and their respecive averages, maxs, etc, etc. 

In [18]:
sp2.describe().round(2)

Item_Visibility  Item_MRP  Outlet_Size  Item_Outlet_Sales
count          8523.00   8523.00      8523.00            8523.00
mean              0.07    140.99         1.83            2181.29
std               0.05     62.28         0.60            1706.50
min               0.00     31.29         1.00              33.29
25%               0.03     93.83         1.00             834.25
50%               0.05    143.01         2.00            1794.33
75%               0.09    185.64         2.00            3101.30
max               0.33    266.89         3.00           13086.96

##Exploratory Visuals

## Explanatory Visuals